File: Daily Questions/0386-lexicographical-numbers/0386-lexicographical-numbers.py
from typing import List

class Solution:
    def lexicalOrder(self, n: int) -> List[int]:
        """
        Generates numbers from 1 to n in lexicographical order using an iterative approach.
        
        Parameters:
        n (int): The upper limit of numbers to be included in lexicographical order.

        Returns:
        List[int]: A list of integers from 1 to n in lexicographical order.
        """
        result = []
        current = 1
        
        for _ in range(n):
            # Append the current number to the result list
            result.append(current)
            
            # Try to go to the next lexicographical "child" by multiplying by 10
            if current * 10 <= n:
                current *= 10
            else:
                # If no child is possible, try the next sibling by incrementing
                if current >= n:
                    current //= 10  # Move up to a level where incrementing is possible
                current += 1
                
                # Remove trailing 9s to ensure proper lexicographical order
                while current % 10 == 0:
                    current //= 10
        
        return result

File: 0386-lexicographical-numbers/0386-lexicographical-numbers.c
#include <stdlib.h>

/**
 * @brief Generates numbers from 1 to n in lexicographical order.
 *
 * This function uses an iterative approach to simulate lexicographical order,
 * starting with 1 and moving to each next lexicographical number until reaching n.
 *
 * @param n (int): The upper limit of numbers to include in lexicographical order.
 * @param returnSize (int*): Pointer to store the size of the returned array.
 *
 * @return (int*): Pointer to an array of integers in lexicographical order.
 *                 The array is dynamically allocated and must be freed by the caller.
 */
int* lexicalOrder(int n, int* returnSize) {
    // Allocate memory for the result array
    int* result = (int*)malloc(n * sizeof(int));
    if (result == NULL) {
        *returnSize = 0;  // In case of memory allocation failure
        return NULL;
    }
    
    *returnSize = n;  // Set the return size to n, as we'll fill it up to n elements
    int current = 1;  // Start with the first lexicographical number

    // Fill the result array with numbers from 1 to n in lexicographical order
    for (int i = 0; i < n; ++i) {
        result[i] = current;  // Add the current number to the result array

        // Move to the next lexicographical number
        if (current * 10 <= n) {
            // Try to move to the next "child" by multiplying by 10
            current *= 10;
        } else {
            // If no child is possible, increment to the next "sibling"
            if (current >= n) {
                // If current exceeds n, move back to a parent level
                current /= 10;
            }
            current += 1;  // Move to the next sibling
            
            // Remove trailing zeros (moving back to a valid lexicographical parent)
            while (current % 10 == 0) {
                current /= 10;
            }
        }
    }
    
    return result;  // Return the populated result array
}
